### DISCUSSION TITLE
Is it possible to use data from an asynchronous API call in router.options.ts?

### DISCUSSION THREAD

**Discussion by gerbenvandijk** [nuxt/nuxt#23300] [Questions]
Hi all,

Currently I'm working on setting up a multi-tenant app, where each tenant can choose between several sets of page templates for the page routes. 

I have setup a test case for this using a composable that returns the folder name of the page set, and is then used to show the correct page templates for each route. This is achieved using the below.

composables/tenantConfig:
```ts
const tenantConfigs = {
    'localhost:3000': {
        pageSet: 'pageSetA'
    },
    'domain1.domain.dev': {
        pageSet: 'pageSetB'
    }
}

export const getHost = () => useState<string>('host', () => getHostFromHeaders().value)

export const getHostFromHeaders = (fallback = 'fallbackhost') => {
    const host = ref(fallback)
    if (process.server) {
        const hostFromHeaders = useRequestHeaders().host
        if (hostFromHeaders) {
            host.value = hostFromHeaders
        }
    }
    return host
}

export const tenantConfig = () => useState<Object>('tenantConfig', () => getCurrentTenantConfig().value)

export const getCurrentTenantConfig = (fallback = tenantConfigs['localhost:3000']) => {
    const tenantConfig = ref(fallback)
    if (process.server) {
        const hostFromHeaders = useRequestHeaders().host
        if (hostFromHeaders) {
            tenantConfig.value = tenantConfigs[hostFromHeaders as keyof typeof tenantConfigs]
        }
    }
    return tenantConfig
}
```

app/router.options.ts:
```ts
import type { RouterConfig } from '@nuxt/schema'
import { tenantConfig } from '~/composables/tenantConfig'
export default <RouterConfig>{
    routes: (_routes) => {
        let pagesDir = 'pageSetA'
        if (tenantConfig().value.themeName) {
            pagesDir = tenantConfig().value.pageSet
        }
        const routes = _routes.filter((route) => {
            return (
                typeof route.name === 'string' && route.name.includes(pagesDir)
            )
        })
        _routes.forEach((route) => {
            route.path = route.path.replace(`/${pagesDir}`, '')
        })
        return routes
    }
}

```
So the above gets the "host" header, matches it with a hardcoded list of hosts/pageSet combos, and returns the correct pageSet and serves the correct page for each route.

That all works great, it is tested on multiple hosts all pointing back to the same application. 

But it depends on a hardcoded list of all tenants and their pageSets. This data has to come from an API eventually, so for the past day I have been trying to implement an API call in router.options.ts but to no avail. 

Is it possible at all to do an async/await in router.options.ts, so I can use the data from an API call to determine what page folder to match with routes? 

I've tried many ways to achieve this but I'm about to conclude that it is not possible by design, even though my test case above works. Is it correct that this is not going to work because of the place of router.options.ts in the lifecycle of the application (is it running on build time and not runtime)?

It's confusing me a bit since I am able to set the folder dynamically based on the "host" header, so why would it not be possible using the response of an API call?

I have also tried different methods, such as using a module with extendPages, using props on NuxtPage component inside app.vue, all to no avail.

I would love to avoid having to conditionals/dynamic components in the rest of the application and be able to abstract this somehow.
Any tips / insights / further resources would be greatly appreciated 👍 💯 

Thanks for the help!

**Comment by or2e** [comment]
[gerbenvandijk](https://github.com/gerbenvandijk)

One solution for you may be to use [addRoute ](https://github.com/nuxt/nuxt/blob/main/test/fixtures/basic/plugins/add-route.ts) and middleware/[plugin](https://github.com/nuxt/nuxt/blob/main/test/fixtures/basic/plugins/add-route.ts)
